File: data-structure-and-algorithm/java/serialize-and-deserialize-binary-tree.java
/**
 * Serialize and Deserialize Binary Tree
 * 
 * Serialization is the process of converting a data structure or object into a sequence of bits so that it can be stored in a file or memory buffer,
 * or transmitted across a network connection link to be reconstructed later in the same or another computer environment.
 * 
 * Design an algorithm to serialize and deserialize a binary tree. There is no restriction on how your serialization/deserialization algorithm should work.
 * You just need to ensure that a binary tree can be serialized to a string and this string can be deserialized to the original tree structure.
 * 
 * For example, you may serialize the following tree
 * 
 *     1
 *    / \
 *   2   3
 *      / \
 *     4   5
 * as "[1,2,3,null,null,4,5]", just the same as how LeetCode OJ serializes a binary tree. You do not necessarily need to follow this format, so please be creative and come up with different approaches yourself.
 * 
 * Note: Do not use class member/global/static variables to store states. Your serialize and deserialize algorithms should be stateless.
 */

public class Codec {
    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder stringBuilder = new StringBuilder();
        serializeHelper(root, stringBuiler);
        return stringBuilder.toString().trim();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        int[] index = new int[0];
        return deserializeHelper(data.replace(" ", "").toCharArray(), index);
    }
    
    private void serializeHelper(TreeNode root, StringBuilder stringBuilder) {
        if (root == null) {
            stringBuilder.append("# ");
        } else {
            stringBuilder.append(root.val)
                .append(" ");
            serializeHelper(root.left, stringBuilder);
            serializeHelper(root.right, stringBuilder);
        }
    }

    private TreeNode deserializeHelper(char[] dataCharArray, int[] index) {
        if (dataCharArray[index[0]] == '#') {
            return null;
        }
        TreeNode node = new TreeNode();
        node.val = Integer.parseInt(new String(dataCharArray[index[0]]));

        index[0] += 1;
        node.left = deserializeHelper(dataCharArray, index);
        index[0] += 1;
        node.right = deserializeHelper(dataCharArray, index);

        return node;
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
        val = 0;
        left = null;
        right = null;
    }
}

File: data-structure-and-algorithm/typescript/leetcode/src/serialize-and-deserialize-binary-tree.ts
/*
Serialization is the process of converting a data structure or object into a sequence of bits so that it can be stored in a file or memory buffer, or transmitted across a network connection link to be reconstructed later in the same or another computer environment.

Design an algorithm to serialize and deserialize a binary tree. There is no restriction on how your serialization/deserialization algorithm should work. You just need to ensure that a binary tree can be serialized to a string and this string can be deserialized to the original tree structure.

Clarification: The input/output format is the same as how LeetCode serializes a binary tree. You do not necessarily need to follow this format, so please be creative and come up with different approaches yourself.



Example 1:


Input: root = [1,2,3,null,null,4,5]
Output: [1,2,3,null,null,4,5]
Example 2:

Input: root = []
Output: []
Example 3:

Input: root = [1]
Output: [1]
Example 4:

Input: root = [1,2]
Output: [1,2]


Constraints:

The number of nodes in the tree is in the range [0, 104].
-1000 <= Node.val <= 1000
*/

import {TreeNode} from './data-structure/BinaryTree';
import {Nullable} from './util/object';

function serialize(root: Nullable<TreeNode<number>>): string {
  const result: (number | string)[] = [];
  serializeDFS(root);
  return result.join(' ');

  function serializeDFS(node: Nullable<TreeNode<number>>): void {
    if (node === null) {
      result.push('#');
    } else {
      result.push(node.data);
      serializeDFS(node.left);
      serializeDFS(node.right);
    }
  }
}

function deserialize(data: string): Nullable<TreeNode<number>> {
  const list = data.split(' ');
  const iter = list.values();
  return deserializeDfs();

  function deserializeDfs(): Nullable<TreeNode<number>> {
    const {value: data, done} = iter.next();
    if (done || data === '#') {
      return null;
    }

    const node = new TreeNode(Number(data));
    node.left = deserializeDfs();
    node.right = deserializeDfs();
    return node;
  }
}
